src/hcluster.js

```diff
@@ -122,20 +125,28 @@ export function hcluster() {
           case 'average':
             distMatrix[j][c1] = distMatrix[c1][j] =
               (cSize[c1] * distMatrix[c1][j] + cSize[c2] * distMatrix[c2][j]) /
-              (cSize[c1] + cSize[j]);
+              (cSize[c1] + cSize[c2]);
             break;
         }
       }
-      distMatrix[c1][c1] = Infinity;
+      cSize[c1] += cSize[c2];
 
       for (let i = 0; i < n; i++) {
         distMatrix[i][c2] = distMatrix[c2][i] = Infinity;
       }
 
       // update dmin and replace ones that previous pointed to c2 to point to c1
       for (let j = 0; j < n; j++) {
-        if (dMin[j] == c2) {
```

Comment:
Maybe the test can be done outside of the loop to avoid testing many times and allow the JIT compiler optimize the loop in both cases?